File: C++/Day04.cpp
#include "AH.h"

namespace Day04
{

	void parseLine(const std::string game, std::vector<int> & winning,
		std::vector<int> & scratchcard)
	{
		winning.clear();
		scratchcard.clear();

		const int space = game.find(":");
		std::string parsed = game.substr(space+1);
		const auto s = AH::Split(parsed, '|');

		const auto wins = AH::Split(AH::trim(s[0]), ' ');
		for (auto & win : wins) {
			if (win.size() == 0) { continue; }
			winning.push_back(stoi(AH::trim(win)));
		}
		const auto cards = AH::Split(AH::trim(s[1]), ' ');
		for (auto & card : cards) {
			if (card.size() == 0) { continue; }
			scratchcard.push_back(stoi(AH::trim(card)));
		}

		return;
	}

	int scoreCard(const std::vector<int> winning, const std::vector<int> scratchcard) {
		int score = 0;
		for (auto card : scratchcard) {
			if (std::find(winning.begin(), winning.end(), card) != winning.end()) {
			  score++;
			}
		}

		return score;
	}

	int Run(const std::string& filename)
	{
		const std::vector<std::string> inputLines = AH::ReadTextFile(filename);
		std::vector<int> winning(0);
		std::vector<int> scratchcard(0);

		std::vector<int> copies(inputLines.size(), 1);
		int part1 = 0, part2 = 0;

		for (size_t i = 0; i < inputLines.size(); ++i) {
			const auto l = inputLines[i];

			parseLine(l, winning, scratchcard);
			int score = scoreCard(winning, scratchcard);
			if (score > 0) {
				for (int j = 0; j < score; ++j) {
					if (i + j + 1 < inputLines.size()) {
						copies[i + j + 1] += copies[i];
					}
				}
				part1 += (1 << (score - 1));
			}
		}
		for (auto c : copies) {
			part2 += c;
		}

		AH::PrintSoln(4, part1, part2);

		return 0;
	}

}

File: Go/Day04/Day04.go
//go:build d04
// +build d04

package Day04

import (
	"slices"
	"strconv"
	"strings"

	AH "AoC/adventhelper"
)

func parseLine(game string) ([]int, []int) {
	winning := []int{}
	scratchcard := []int{}
	parts := strings.Split(game, ":")
	s := strings.Split(parts[1], "|")

	wins := strings.Split(s[0], " ")
	for _, win := range wins {
		if len(win) == 0 {
			continue
		}
		win_val, _ := strconv.Atoi(win)
		winning = append(winning, win_val)
	}

	cards := strings.Split(s[1], " ")
	for _, card := range cards {
		if len(card) == 0 {
			continue
		}
		card_val, _ := strconv.Atoi(card)
		scratchcard = append(scratchcard, card_val)
	}

	return winning, scratchcard
}

func scoreCard(winning []int, card []int) (score int) {
	score = 0
	for _, number := range card {
		if slices.Contains(winning, number) {
			score++
		}
	}

	return
}

func Run() {
	inputLines, _ := AH.ReadStrFile("../input/input04.txt")

	part1, part2 := 0, 0
	copies := make([]int, len(inputLines))
	for i := range copies {
		copies[i] = 1
	}
	for i, l := range inputLines {
		winning, scratchcard := parseLine(l)
		score := scoreCard(winning, scratchcard)
		if score > 0 {
			part1 += (1 << (score - 1))

			for j := 0; j < score; j++ {
				if (i + j + 1) < len(copies) {
					copies[i+j+1] += copies[i]
				}
			}
		}
	}

	for _, c := range copies {
		part2 += c
	}

	AH.PrintSoln(4, part1, part2)

	return
}
